Documentation/teaching/lectures/interrupts.rst

```diff
@@ -0,0 +1,558 @@
+                               |
+                      +--------+--------+
+                      |                 |
+                      |    I/O APIC     |
+                      |                 |
+                      +--------+--------+
+                               |
+                               |
+                               |
+                      External interrupts
+
+With multicore systems, each core has a local APIC used to process interrupts
+from locally connected devices like timers or thermals sensors.
+
+I/O APIC is used to distributed IRQ from external devices to CPU cores.
+
+After discussing the hardware, now let's see how the processor handles an interrupt.
+
+Interrupt Descriptor Table
+--------------------------
+
+The interrupt descriptor table (IDT) associates each interrupt or exception
+identifier with a descriptor for the instructions that service the associated
+event. We will name the identifier as vector number and the associated 
+instructions as interrupt/exception handler.
+
+An, IDT thus is:
+
+.. slide:: Interrupt Descriptor Table
+   :inline-contents: True
+   :level: 2
+
+   * used as a jump table by the CPU when a given vector is triggered
+   * array of 256 x 8 bytes entries
+   * may reside anywhere in physical memory
+   * processor locates IDT by the means of IDTR
+
+Below we can find Linux IRQ vector layout. The first 32 entries are reserved
+for exceptions, vector 128 is used for sycall interface and the rest are
+used mostly for hardware interrupts handlers.
+
+.. slide:: Linux IRQ vector layout
+   :inline-contents: True
+   :level: 2
+
+   .. ditaa::
+
+    arch/x86/include/asm/irq_vectors.h
+         +------+
+         |  0   | 0..31, system traps and exceptions
+         +------+
+         |  1   |
+         +------+
+         |      |
+         +------+
+         |      |
+         |      |
+         |      |
+         +------+
+         |  32  |  32..127, device interrupts
+         +------+
+         |      |
+         |      |
+         |      |
+         +------+
+         | 128  |  int80 syscall interface
+         +------+
+         | 129  |  129..255, other interrupts
+         +------+
+         |      |
+         |      |
+         |      |
+         +------+
+         | 255  |
+         +------+
+
+On x86 an IDT entry has 8 bytes and it is named gate. There can be 3 types of gates:
+
+  * interrupt gate, holds the address of an interupt or exception handler.
+    Jumping to the handler disables maskable interrupts (IF flag is cleared).
+  * trap gates, similar with an interrupt gate but it does not disable maskable
+    interrupts while jumping to interupt/exception handler.
+  * task gates (not used in Linux)
+
+Lets have a look at several fields of an IDT entry:
+
+   * segment selector, index into GDT/LDT to find the start of the code segment where
+     the interupt handlers resides
+   * offset, offset inside the code segment
+   * T, represents the type of gate
+
```

Comment:
Maybe add a short description for DPL as well?

Reply:
ok.